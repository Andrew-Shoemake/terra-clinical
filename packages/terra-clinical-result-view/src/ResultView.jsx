import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import IconDocuments from 'terra-icon/lib/icon/IconDocuments';
import IconComment from 'terra-icon/lib/icon/IconComment';
import IconModified from 'terra-icon/lib/icon/IconModified';
import IconUnexpected from 'terra-icon/lib/icon/IconUnexpected';

import IconAbnormal from 'terra-icon/lib/icon/IconAbnormal';
import IconAlert from 'terra-icon/lib/icon/IconAlert';
import IconHigh from 'terra-icon/lib/icon/IconHigh';
import IconLow from 'terra-icon/lib/icon/IconLow';
import styles from './ResultView.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   *  An array of one or more results.
   */
  results: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.number,
    units: PropTypes.string,
    normalcy: PropTypes.string,
  })),
  /**
   *  The age display of the result. Does not handle formatting.
   */
  timeDisplay: PropTypes.string,
  /**
   *  Does the result have related documents. Adds an indicator.
   */
  hasDocument: PropTypes.bool,
  /**
   *  Has the result been modified. Adds an indicator.
   */
  isModified: PropTypes.bool,
  /**
   *  Are there comments related to the result. Adds an indicator.
   */
  hasComment: PropTypes.bool,
  /**
   *  Is the view aligned to the right. Defaults to false.
   */
  alignEnd: PropTypes.bool,
  /**
   *  Can the result text be truncated. Defaults to true.
   */
  isTruncated: PropTypes.bool,
  /**
   *  Should the view have padding. Defaults to false.
   */
  isPadded: PropTypes.bool,
};

const defaultProps = {
  results: [],
  timeDisplay: null,
  hasDocument: false,
  hasComment: false,
  isModified: false,
  alignEnd: false,
  isTruncated: true,
  isPadded: false,
};

const normalcyIndicators = {
  ABNORMAL: <IconAbnormal />,
  CRITICAL: <IconAlert />,
  LOW: <IconLow />,
  HIGH: <IconHigh />,
};

const resultIndicator = (result) => {
  let icon = null;

  if (!result.normalcy || !result.normalcy.length) {
    return null;
  } else if (Object.keys(normalcyIndicators).includes(result.normalcy)) {
    icon = normalcyIndicators[result.normalcy];
  }

  return <span className={cx('result-view-indicator', result.normalcy.toLowerCase())} >{icon}</span>;
};

const resultsDiv = (results, isTruncated) => {
  if (!results || !results.length) {
    const noValueDisplay = '--';
    return <div className={cx('result-view-result-value-text')} >{noValueDisplay}</div>;
  }

  const resultsDivs = [];

  for (let i = 0; i < results.length; i += 1) {
    if (resultsDivs.length >= 1) {
      const separator = <span className={cx('result-view-result-separator')} >{'/'}</span>;
      resultsDivs.push(separator);
    }

    const resultValue = results[i];
    const resultIndicatorDiv = resultIndicator(resultValue);

    if (resultIndicatorDiv) {
      resultsDivs.push(resultIndicatorDiv);
    }

    if (resultValue.unexpected) {
      resultsDivs.push(<IconUnexpected />);
    }

    let unitDiv = null;
    if (resultValue.unit && resultValue.unit.length > 0) {
      unitDiv = <small className={cx('result-view-result-unit')} >{resultValue.unit}</small>;
    }

    let resultValueDisplay = '--';
    if (resultValue.value && resultValue.value.length > 0) {
      resultValueDisplay = resultValue.value;
    }

    let resultValueTextClassName = 'result-view-result-value-text';
    if (isTruncated) {
      resultValueTextClassName = 'result-view-is-truncated';
    }

    let normalcyLevel = results[i].normalcy;
    if (normalcyLevel !== undefined) {
      normalcyLevel = normalcyLevel.toLowerCase();
    }

    const resultValueDiv = (<span className={cx(resultValueTextClassName, normalcyLevel)} >
      {resultValueDisplay}
      {unitDiv}
    </span>);

    resultsDivs.push(resultValueDiv);
  }

  return resultsDivs;
};

const timeDisplayDiv = (timeDisplay) => {
  const resultAgeClassName = 'result-view-result-age';
  return <div className={cx(resultAgeClassName)} >{timeDisplay}</div>;
};

const ResultView = (props) => {
  const { results, timeDisplay, hasDocument, isModified, hasComment, alignEnd, isTruncated, isPadded } = props;
  const attributesName = 'result-view';
  let align = 'align-start';

  if (alignEnd) {
    align = 'align-end';
  }

  let bodyClassName = 'result-view-body';

  if (isPadded) {
    bodyClassName = 'result-view-body--padded';
  }

  return (
    <div className={cx(attributesName, align)} >
      <div className={cx(bodyClassName)}>
        <div className={cx('result-view-result-block')}>
          {hasDocument &&
            <IconDocuments />
          }
          { resultsDiv(results, isTruncated) }
          {isModified &&
            <IconModified />
          }
          {hasComment &&
            <IconComment />
          }
        </div>
        {timeDisplay &&
          timeDisplayDiv(timeDisplay)
        }
      </div>
    </div>
  );
};

ResultView.defaultProps = defaultProps;
ResultView.propTypes = propTypes;

export default ResultView;
export { normalcyIndicators };
