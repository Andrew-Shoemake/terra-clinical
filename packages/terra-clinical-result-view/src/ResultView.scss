@import './mixins';

:local {
  .result-view {
    color: var(--result-view-background-color, #595f62);
    font-size: var(--result-view-font-size, 20px);
    line-height: var(--result-view-line-height, 20px);
    white-space: nowrap;
    width: 100%;

    svg {
      fill: currentColor;
    }

    &.alignment-left {
      text-align: left;
    }

    &.alignment-right {
      text-align: right;
    }

    &.alignment-center {
      text-align: center;
    }
  }

  .result-view-body--padded {
    padding: var(--result-view-body-padding, 5px);
  }

  .result-view-result-block {
    align-items: baseline;
    display: inline-flex;
  }

  .result-view-indicator {
    padding-right: var(--result-view-indicator-padding-right, 3px);
  }

  .result-view-result-value-text {
    font-size: var(--result-view-result-value-font-size, 20px);
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .result-view-result-unit {
    color: var(--result-view-result-unit-text-color, #6f7477);
    line-height: normal;
    padding-left: var(--result-view-result-unit-padding-left, 3px);
    padding-right: var(--result-view-result-unit-padding-right, 3px);
  }

  .result-view-result-age {
    color: var(--result-view-result-age-text-color, #6f7477);
    display: block;
    font-size: var(--result-view-result-age-font-size, 20px);
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .result-view-truncated {
    @include terra-clinical-text-truncate;
  }

  .result-view-result-separator {
    padding-left: var(--result-view-result-separator-padding-left, 3px);
    padding-right: var(--result-view-result-separator-padding-right, 3px);
  }
}
