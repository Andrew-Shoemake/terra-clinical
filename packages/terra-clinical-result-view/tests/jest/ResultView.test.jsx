import React from 'react';
import ResultView from '../../src/ResultView';

describe('ResultView', () => {
  const defaultRender = <ResultView />;

  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Prop Tests
  it('should use the default value when no value is given', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.find('.result-view').text()).toEqual('default');
  });

  // Structure Tests
  it('should have the class result-view', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.prop('className')).toContain('result-view');
  });
});
