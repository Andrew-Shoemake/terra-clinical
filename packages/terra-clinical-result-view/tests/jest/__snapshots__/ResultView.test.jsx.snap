// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultView should render a Result View with RTL support 1`] = `
<ResultView
  alignEnd={true}
  alignment="left"
  hasComment={true}
  hasDocument={true}
  isModified={false}
  isPadded={true}
  isTruncated={false}
  results={
    Array [
      Object {
        "color": "#0053e6",
        "icon": <IconLow
          id="Layer_1"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        />,
        "unit": "ml",
        "value": 503456789,
      },
    ]
  }
  timeDisplay={null}
/>
`;

exports[`ResultView should render a Result View with no time display 1`] = `
<ResultView
  alignment="left"
  hasComment={true}
  hasDocument={true}
  isModified={false}
  isPadded={true}
  isTruncated={true}
  results={
    Array [
      Object {
        "color": "#0053e6",
        "icon": <IconLow
          id="Layer_1"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        />,
        "unit": "ml",
        "value": 503456789,
      },
    ]
  }
  timeDisplay={null}
/>
`;

exports[`ResultView should render a Result View with separator 1`] = `
<ResultView
  alignment="left"
  hasComment={true}
  hasDocument={true}
  isModified={false}
  isPadded={true}
  isTruncated={false}
  results={
    Array [
      Object {
        "unit": "mmHg",
        "value": 120,
      },
      Object {
        "unit": "mmHg",
        "value": 80,
      },
    ]
  }
  timeDisplay={null}
/>
`;

exports[`ResultView should render a Result View with truncation by default 1`] = `
<ResultView
  alignment="left"
  hasComment={true}
  hasDocument={true}
  isModified={false}
  isPadded={true}
  isTruncated={true}
  results={
    Array [
      Object {
        "color": "#0053e6",
        "icon": <IconLow
          id="Layer_1"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        />,
        "unit": "ml",
        "value": 503456789,
      },
    ]
  }
  timeDisplay={null}
/>
`;

exports[`ResultView should render a Result View without truncation 1`] = `
<ResultView
  alignment="left"
  hasComment={true}
  hasDocument={true}
  isModified={false}
  isPadded={true}
  isTruncated={false}
  results={
    Array [
      Object {
        "color": "#0053e6",
        "icon": <IconLow
          id="Layer_1"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        />,
        "unit": "ml",
        "value": 503456789,
      },
    ]
  }
  timeDisplay={null}
/>
`;

exports[`ResultView should render a default component 1`] = `
<ResultView
  alignment="left"
  hasComment={false}
  hasDocument={false}
  isModified={false}
  isPadded={false}
  isTruncated={true}
  results={Array []}
  timeDisplay={null}
/>
`;
